codex-rs/core/src/config.rs

```diff
@@ -340,6 +343,129 @@ pub fn set_project_trusted(codex_home: &Path, project_path: &Path) -> anyhow::Re
     Ok(())
 }
 
+/// Persist the default `model` and `model_reasoning_effort` to
+/// `CODEX_HOME/config.toml` so the selection is used across sessions.
+///
+/// If a `profile` is set in `config.toml`, this updates the corresponding
+/// `[profiles.<name>]` table; otherwise it updates the top-level keys.
+pub fn set_default_model_and_effort(
+    codex_home: &Path,
+    model: &str,
+    effort: ReasoningEffort,
+) -> anyhow::Result<()> {
+    set_default_model_and_effort_for_profile(codex_home, None, model, effort)
+}
+
+/// Persist defaults under the specified profile if provided; otherwise, if a
+/// `profile` is set in `config.toml`, use it; if neither is present, update
+/// the top-level keys.
+pub fn set_default_model_and_effort_for_profile(
+    codex_home: &Path,
+    profile_override: Option<&str>,
+    model: &str,
+    effort: ReasoningEffort,
+) -> anyhow::Result<()> {
+    let effort_str = effort.to_string();
+    let overrides: [(&[&str], &str); 2] = [
+        (&["model"], model),
+        (&["model_reasoning_effort"], effort_str.as_str()),
+    ];
+    persist_overrides(codex_home, profile_override, &overrides)
+}
+
+/// Persist overrides into `config.toml` using explicit key segments per
+/// override. This avoids ambiguity with keys that contain dots or spaces.
+fn persist_overrides(
+    codex_home: &Path,
+    profile: Option<&str>,
+    overrides: &[(&[&str], &str)],
+) -> anyhow::Result<()> {
+    let config_path = codex_home.join(CONFIG_TOML_FILE);
+
+    let mut doc = match std::fs::read_to_string(&config_path) {
+        Ok(s) => s.parse::<DocumentMut>()?,
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => DocumentMut::new(),
+        Err(e) => return Err(e.into()),
+    };
+
+    let effective_profile: Option<String> = match profile {
+        Some(name) => Some(name.to_string()),
+        None => load_config_as_toml(codex_home).ok().and_then(|v| {
+            v.get("profile")
+                .and_then(|i| i.as_str())
+                .map(|s| s.to_string())
+        }),
+    };
```

Comment:
I asked chat to turn this into a one-liner, but I haven't verified this compiles:

```suggestion
let effective_profile = profile
    .as_deref()
    .or_else(|| {
        load_config_as_toml(codex_home)
            .ok()
            .and_then(|v| v.get("profile").and_then(|i| i.as_str()))
    })
    .map(ToString::to_string);
```